climada/hazard/base.py

```diff
@@ -1396,47 +1394,99 @@ def _read_att_excel(self, file_name, var_names):
+        for haz in haz_list:
+            haz._check_events()
+
+        # check type, unit, and attribute consistency among hazards
+        haz_types = {haz.tag.haz_type for haz in haz_list if haz.tag.haz_type != ''}
+        if len(haz_types) > 1:
+            raise ValueError(f"The given hazards are of different types: {haz_types}. "
+                             "The hazards are incompatible and cannot be concatenated.")
+        self.tag.haz_type = haz_types.pop()
+
+        haz_classes = {type(haz) for haz in haz_list}
+        if len(haz_classes) > 1:
+            raise TypeError(f"The given hazards are of different classes: {haz_classes}. "
+                            "The hazards are incompatible and cannot be concatenated.")
+
+        units = {haz.units for haz in haz_list if haz.units != ''}
+        if len(units) > 1:
+            raise ValueError(f"The hazards use different units: {units}. "
+                             "The hazards are incompatible and cannot be concatenated.")
+        elif len(units) == 0:
+            units = {''}
+        self.units = units.pop()
+
+        attributes = sorted(set.union(*[set(vars(haz).keys()) for haz in haz_list]))
+        for attr_name in attributes:
+            if not all(hasattr(haz, attr_name) for haz in haz_list_nonempty):
+                raise ValueError(f"Attribute {attr_name} is not shared by all hazards. "
+                                 "The hazards are incompatible and cannot be concatenated.")
+
+        # append all tags (to keep track of input files and descriptions)
+        for haz in haz_list:
+            if haz.tag is not self.tag:
+                self.tag.append(haz.tag)
+
+        # map individual centroids objects to union
+        centroids = Centroids.union(*[haz.centroids for haz in haz_list])
+        hazcent_in_cent_idx_list = [
+            u_coord.assign_coordinates(haz.centroids.coord, centroids.coord, threshold=0)
+            for haz in haz_list_nonempty
+            ]
+
+        # concatenate array and list attributes of non-empty hazards
+        for attr_name in attributes:
+            attr_val_list = [getattr(haz, attr_name) for haz in haz_list_nonempty]
+            if isinstance(attr_val_list[0], sparse.csr.csr_matrix):
+                # map sparse matrix onto centroids
+                setattr(self, attr_name, sparse.vstack([
+                    sparse.csr_matrix(
+                        (matrix.data, cent_idx[matrix.indices], matrix.indptr),
+                        shape=(matrix.shape[0], centroids.size)
+                        )
+                    for matrix, cent_idx in zip(attr_val_list, hazcent_in_cent_idx_list)
+                    ], format='csr'))
+            elif isinstance(attr_val_list[0], np.ndarray) and attr_val_list[0].ndim == 1:
+                setattr(self, attr_name, np.hstack(attr_val_list))
+            elif isinstance(attr_val_list[0], list):
+                setattr(self, attr_name, sum(attr_val_list, []))
```

Comment:
There we must think a bit about what to do with non-standard parameters. For instance `WildFire` objects have attributes of the custom dataclasses `FirmsParams()` and `ProbaParams()`.

Reply:
In my opinion, somebody who introduces the non-standard parameters must deal with this. If there is absolutely no way to adapt the non-standard parameters to fit to the current implementation of the concatenation, then that developer can create a new PR and suggest changes. But we can't a priori have support for all kinds of non-standard parameters. By supporting lists we are extremely flexible already. Other objects might have Pandas Series or all kinds of list-like objects in the future. But I suggest to support this, only as soon as somebody asks for it.

Reply:
In the meantime, I had a look at the `WildFire` class. Since `FirmsParams` and `ProbaParams` aren't iterables and don't depend on the event size, your comment doesn't seem to be related to this part of the code, but is of a more general concern.

When subclasses add attributes that are not event-specific, but ensemble-specific. It's still up to the creators of those subclasses to make sure that they work as expected with `append`. Currently, since the append-operation is in-place, the attributes of the `self` object remain untouched unless they are event-specific (list/array) or tag attributes.

A more complex approach would be for subclasses to define "compatibility attributes": a class-specific set of attributes that need to be identical in order for objects to be appendable. For the base class, those attributes would be `haz_type` and `units`. Subclasses would extend that set similar to the current use of the `vars_opt` class attribute.

Reply:
>When subclasses add attributes that are not event-specific, but ensemble-specific, it's still up to the creators of those subclasses to make sure that they work as expected with append

I do not think this is a feasible requirement. How should a developer in the future that when developing a Hazard they should specifically go look into the code of the `hazard.append` method to ensure compatibility?

Reply:
Yes, that's how inheritance works. This is nothing that we "require", but it's just a fact. When you inherit from a class that somebody else defines, then you yourself have to make sure that you don't break the base class's functionality. It's always like that. There is no alternative.

Reply:
I agree with the principle. But if the issue is hidden in the code, that becomes increasingly difficult.